File: app/src/main/java/com/example/myapp/DBService/ShamanDao.java
package com.example.myapp.DBService;

import androidx.room.Dao;
import androidx.room.Delete;
import androidx.room.Insert;
import androidx.room.OnConflictStrategy;
import androidx.room.Query;
import androidx.room.Update;

import java.util.List;

@Dao
public interface ShamanDao {

    @Insert(onConflict = OnConflictStrategy.REPLACE)
    void insertLocation(Location location);

    @Update(onConflict = OnConflictStrategy.REPLACE)
    void updateLocation(Location location);

    @Delete
    void deleteLocation(Location location);

    @Query("DELETE FROM locations WHERE id = :id")
    void deleteLocationById(long id);

    @Query("SELECT * FROM locations WHERE favorite = 1")
    List<Location> getFavoriteLocations();

    @Insert(onConflict = OnConflictStrategy.ABORT)
    void insertRequest(Request request);

    @Delete
    void deleteRequest(Request request);

    @Query("DELETE FROM requests WHERE time = :time")
    void deleteRequestByTime(long time);

    @Query("SELECT l.name, l.country, r.time, r.temperature, l.favorite FROM requests AS r JOIN locations AS l ON l.id = r.location_id")
    List<RequestForAll> getAllRequests();

    @Query("SELECT time, temperature FROM requests WHERE location_id = :locationId")
    List<RequestForLocation> getLocationRequests(long locationId);
}

File: app/src/main/java/com/chplalex/shaman/DBService/ShamanDao.kt
package com.chplalex.shaman.DBService

import androidx.room.Dao
import androidx.room.Delete
import androidx.room.Insert
import androidx.room.OnConflictStrategy
import androidx.room.Query
import androidx.room.Update

@Dao interface ShamanDao {

    @Insert(onConflict = OnConflictStrategy.IGNORE)
    fun insertLocation(location: Location)

    @Update(onConflict = OnConflictStrategy.REPLACE)
    fun updateLocation(location: Location)

    @Query("UPDATE locations SET favorite = :favorite WHERE name = :name AND country = :country")
    fun updateLocationFavoriteByNameAndCountry(name: String, country: String, favorite: Boolean)

    @Delete
    fun deleteLocation(location: Location)

    @Query("DELETE FROM locations WHERE id = :id")
    fun deleteLocationById(id: Long)

    @get:Query("SELECT * FROM locations WHERE favorite = 1")
    val favoriteLocations: MutableList<Location>

    @Query("SELECT * FROM locations WHERE name = :name AND country = :country")
    fun getLocationByNameAndCountry(name: String, country: String): List<Location>

    @Insert(onConflict = OnConflictStrategy.ABORT)
    fun insertRequest(request: Request)

    @Delete
    fun deleteRequest(request: Request)

    @Query("DELETE FROM requests WHERE time = :time")
    fun deleteRequestByTime(time: Long)

    @get:Query("SELECT l.name, l.country, r.time, r.temperature, l.favorite FROM requests AS r JOIN locations AS l ON l.id = r.location_id ORDER BY r.time DESC")
    val allRequests: MutableList<RequestForAll>

    @Query("SELECT time, temperature FROM requests WHERE location_id = :locationId")
    fun getLocationRequests(locationId: Long): List<RequestForLocation>
}